skhep/math/vectors.py

```diff
@@ -3,1029 +3,394 @@
-        self.setpxpypzm(px,py,pz,m)
-        
-    def setptetaphie(self, pt, eta, phi, e):
-        """ Set the transverse momentum, the pseudorapidity, the angle phi and the energy."""
-        px, py, pz = pt*cos(phi), pt*sin(phi), pt*sinh(eta)
-        self.setpxpypze(px,py,pz,e)
-        
-    def tolist(self):
-        """Return the LorentzVector as a list."""
-        return list(self.__vector3d) + [self.__t]
-        
-    def __setitem__(self, i, value):
-        """Update/set the ith vector component (commencing at 0, of course)."""
-        try:
-            if i == 3: self.__t = value
-            else: self.__vector3d[i] = value
-        except IndexError:
-            raise IndexError(
-                'LorentzVector is of length {0} only!'.format(len(self)))
-
-    def __getitem__(self, i):
-        """Get the ith vector component (commencing at 0, of course)."""
-        try:
-            return self.tolist()[i]
-        except IndexError:
-            raise IndexError( 'LorentzVector is of length {0} only!'.format(len(self)))
-
-    def __len__(self):
-        """Length of the LorentzVector, i.e. the number of elements = 4."""
-        return len(self.tolist())
-        
-    @property
-    def p(self):
-        """Return the momentum, aka norm of the momentum vector."""
-        return self.__vector3d.mag
-        
-    @property
-    def pt(self):
-        """Return the transverse momentum, aka transverse component of the momentum vector."""
-        return self.perp
-        
-    @property
-    def et(self):
-        """Return the transverse energy."""
-        return self.e * ( self.pt / self.p )
-
-    @property
-    def m(self):
-        """Return the invariant mass."""
-        return self.mag
-
-    @property
-    def m2(self):
-        """Return the square of the invariant mass."""
-        return self.mag2
-        
-    @property
-    def mass(self):
-        """Return the invariant mass."""
-        return self.mag
+    def rho(self):
+        return self[:2].mag().view(np.ndarray)
+
+    def angle(self, other):
+        'Angle between two vectors'
+        return super(Vector2D, self).angle(other)
+
+class Vector3D(Vector2D):
+    __slots__ = ()
+    NAMES = ('x', 'y', 'z')
+
+    def __new__(cls, x, y, z, dtype=np.double):
+        return Vector.__new__(cls, x, y, z, dtype=dtype)
+
+    def cross(self, other):
+        return Vector3D(self.y*other.z - self.z*other.y,
+                        self.z*other.x - self.x*other.z,
+                        self.x*other.y - self.y*other.x)
+
+    def theta(self):
+        prep = np.sqrt(self.x*self.x + self.y*self.y)
+        return np.arctan2(prep,self.z).view(np.ndarray)
+
+
+    def in_basis(self, xhat, yhat, zhat):
+        '''Must be unit vectors, should be orthogonal'''
+        return Vector3D(self.dot(xhat),
+                        self.dot(yhat),
+                        self.dot(zhat))
+
+    angle = Vector.angle
+
+    def rotate(self, phi=0, theta=0, psi=0):
```

Comment:
Here you use 3 angles to do a rotation, while in the current version a rotation is defined by one angle and one directional angle. Is there a preferred way or should we keep both?

Reply:
ROOT provides both. I was replacing `.rotateX(angle)` with `.rotate(x=angle)`, but since `.rotate` also exists for axis/angle rotation, we need to rename (at least) one of them.